catalog/core/catalog-core-versioning/versioning-common/src/main/java/ddf/catalog/core/versioning/MetacardVersion.java

```diff
@@ -46,134 +59,186 @@
                 true /* indexed */,
                 true /* stored */,
                 false /* tokenized */,
                 false /* multivalued */,
                 BasicTypes.DATE_TYPE));
-        versionDescriptors.add(new AttributeDescriptorImpl(VERSION_OF_ID,
+        VERSION_DESCRIPTORS.add(new AttributeDescriptorImpl(VERSION_OF_ID,
                 true /* indexed */,
                 true /* stored */,
                 false /* tokenized */,
                 false /* multivalued */,
                 BasicTypes.STRING_TYPE));
-        versionDescriptors.add(new AttributeDescriptorImpl(VERSION_TAGS,
+        VERSION_DESCRIPTORS.add(new AttributeDescriptorImpl(VERSION_TAGS,
                 true /* indexed */,
                 true /* stored */,
                 false /* tokenized */,
                 true /* multivalued */,
                 BasicTypes.STRING_TYPE));
+        VERSION_DESCRIPTORS.add(new AttributeDescriptorImpl(VERSION_TYPE,
+                true /* indexed */,
+                true /* stored */,
+                false /* tokenized */,
+                false /* multivalued */,
+                BasicTypes.STRING_TYPE));
+        VERSION_DESCRIPTORS.add(new AttributeDescriptorImpl(VERSION_TYPE_BINARY,
+                false /* indexed */,
+                true /* stored */,
+                false /* tokenized */,
+                false /* multivalued */,
+                BasicTypes.BINARY_TYPE));
+        METACARD_VERSION = new MetacardTypeImpl(VERSION_PREFIX, VERSION_DESCRIPTORS);
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////
+    // ENUMERATIONS
+    ////////////////////////////////////////////////////////////////////////////////////
+    public enum Action {
+        // @formatter:off
+        CREATED("Created"),
+        CREATED_CONTENT("Created-Content"),
+        UPDATED("Updated"),
+        UPDATED_CONTENT("Updated-Content"),
+        DELETED("Deleted");
+        // @formatter:on
+
+        private static Map<String, Action> keyMap = new HashMap<>();
+
+        static {
+            for (Action action : Action.values()) {
+                keyMap.put(action.getKey(), action);
+            }
+        }
+
+        private String key;
+
+        Action(String key) {
+            this.key = key;
+        }
+
+        public String getKey() {
+            return this.key;
+        }
+
+        public static Action fromKey(String key) {
+            return keyMap.get(key);
+        }
+
+        public static Action ofMetacard(Metacard metacard) {
+            if (isNotVersion(metacard)) {
+                throw new IllegalArgumentException(
+                        "Cannot get action of a non version metacard [" + metacard.getId() + "]");
+            }
+            Serializable svalue = Optional.ofNullable(metacard.getAttribute(ACTION))
+                    .map(Attribute::getValue)
+                    .orElse(null);
+            if (!(svalue instanceof String)) {
```

Comment:
Feels like you could probably do your test/cast in the optional processing and return a String. Then if the returned String is null, throw your `IllegalArgumentException`.